packages/react-router-dom/__tests__/special-characters-test.tsx

```diff
@@ -0,0 +1,712 @@
+  },
+
+  // Add a few multi-char space use cases for good measure
+  { char: "a b", pathChar: "a%20b", searchChar: "a%20b", hashChar: "a%20b" },
+  { char: "a+b", pathChar: "a+b", searchChar: "a+b", hashChar: "a+b" },
+];
+
+describe("special character tests", () => {
+  // Mutable vars we'll use to capture the useLocation/useParams values during
+  // the render pass.  this avoids stringifying them into the DOM and parsing
+  // them back out which can mess with the encoding
+  let renderedUseLocation: Omit<Location, "state" | "key"> | null = null;
+  let renderedParams: Params<string> | null = null;
+
+  function CaptureLocation() {
+    let location = {
+      ...useLocation(),
+      state: undefined,
+      key: undefined,
+    };
+    let params = useParams();
+    renderedUseLocation = {
+      pathname: location.pathname,
+      search: location.search,
+      hash: location.hash,
+    };
+    renderedParams = params;
+    return (
+      <>
+        <p>{location.pathname}</p>
+        <Link to="/reset">Link to reset</Link>
+      </>
+    );
+  }
+
+  beforeEach(() => {
+    renderedUseLocation = null;
+    renderedParams = null;
+  });
+
+  describe("when matching as param values", () => {
+    async function testParamValues(
+      navigatePath: string,
+      expectedHeading: string,
+      expectedLocation: Omit<Location, "state" | "key">,
+      expectedParams = {}
+    ) {
+      // Need to use our own custom DOM in order to get a working history
+      const dom = new JSDOM(`<!DOCTYPE html><html><body></body></html>`, {
+        url: "https://remix.run/",
+      });
+      let testWindow = dom.window as unknown as Window;
+      testWindow.history.replaceState(null, "", navigatePath);
+
+      function Comp({ heading }) {
+        return (
+          <>
+            <h1>{heading}</h1>
+            <CaptureLocation></CaptureLocation>
+          </>
+        );
+      }
+
+      let routeElements = (
+        <>
+          <Route path="/path" element={<Comp heading="Static Route" />} />
+          <Route
+            path="/inline-param/:slug"
+            element={<Comp heading="Inline Nested Param Route" />}
+          />
+          <Route path="/param">
+            <Route
+              path=":slug"
+              element={<Comp heading="Parent Nested Param Route" />}
+            />
+          </Route>
+          <Route
+            path="/inline-splat/*"
+            element={<Comp heading="Inline Nested Splat Route" />}
+          />
+          <Route path="/splat">
+            <Route
+              path="*"
+              element={<Comp heading="Parent Nested Splat Route" />}
+            />
+          </Route>
+          <Route
+            path="/reset"
+            element={<Link to={navigatePath}>Link to path</Link>}
+          />
+          <Route path="/*" element={<Comp heading="Root Splat Route" />} />
+        </>
+      );
+
+      // Render BrowserRouter at the initialized location and confirm we get
```

Comment:
The section here and below basically checks 4 things:
* We match routes properly on initial `BrowserRouter` load (reading from `window.location`)
* We match routes properly on client-side `BrowserRouter` navigations (through `history.push`)
* We match routes properly on initial `createBrowserRouter`/`RouterProvider` load (reading from `window.location`)
* We match route properly on client-side `createBrowserRouter`/`RouterProvider` navigations (through `router.navigate`)